src/os/bluestore/bluestore_types.h

```diff
@@ -514,13 +514,22 @@ struct bluestore_blob_t {
     assert(struct_v == 1 || struct_v == 2);
     denc(extents, p);
     denc_varint(flags, p);
-    denc_varint_lowz(logical_length, p);
-    denc_varint_lowz(compressed_length, p);
-    denc(csum_type, p);
-    denc(csum_chunk_order, p);
-    denc_varint(csum_data.length(), p);
-    p += csum_data.length();
-    p += sizeof(unused_t);
+
+    if (is_compressed()) {
+      denc_varint_lowz(logical_length, p);
+      denc_varint_lowz(compressed_length, p);
+    }
+
+    if (has_csum()) {
+      denc(csum_type, p);
+      denc(csum_chunk_order, p);
+      denc_varint(csum_data.length(), p);
+      p += csum_data.length();
+    }
+
+    if (has_unused()) {
```

Comment:
I'm not certain this will be faster.. the value in doing a conservative bound is that it mostly compiles down to nothing.  Everything up until csum_data.length() is a constant and compiles into a single add instruction.  This three conditional branches.  And when we encode these the buffer is temporary (used to write to kv db and then thrown away).